app/models/container_deployment.rb

```diff
@@ -6,5 +6,314 @@ class ContainerDeployment < ApplicationRecord
+    valid_provider = provider.authentication_check.first
+    if valid_provider
+      self.deployed_ems = provider
+      save!
+    else
+      provider.destroy
+    end
+    valid_provider
+  end
+
+  def generate_ansible_inventory_for_subscription
+    template = <<eos
+[OSEv3:children]
+nodes
+#{"masters" unless (roles_addresses("master") - [roles_addresses("deployment_master")]).empty?}
+
+[OSEv3:vars]
+ansible_ssh_user=#{ssh_auth.userid}
+deployment_type=openshift-enterprise
+rhsub_user=#{rhsm_auth.userid}
+rhsub_pass=#{rhsm_auth.password}
+rhsub_pool=#{rhsm_auth.rhsm_sku}
+ansible_sudo=true
+
+[nodes]
+#{(roles_addresses("node") - [roles_addresses("deployment_master")]).join("\n") unless roles_addresses("node").empty?}
+
+#{"[masters]" unless (roles_addresses("master") - [roles_addresses("deployment_master")]).empty?}
+#{(roles_addresses("master") - [roles_addresses("deployment_master")]).join("\n") unless (roles_addresses("master") - [roles_addresses("deployment_master")]).empty?}
+eos
+    template
+  end
+
+  def generate_roles
+    result = {}
+    roles = []
+    container_deployment_nodes.each do |node|
+      roles << node.roles
+    end
+    roles = roles.flatten.uniq
+    result["master"] = {"osm_use_cockpit" => "false"} if roles.include? "master"
+    result["node"] = {} if roles.include? "node"
+    result["storage"] = {} if roles.include? "storage"
+    result
+  end
+
+  def generate_ansible_yaml
+    hash = {}
+    hash["ansible_config"] = ANSIBLE_CONFIG_LOCATION
+    hash["ansible_log_path"] = ANSIBLE_CONFIG_LOG
+    hash["ansible_inventory_path"] = ANSIBLE_CONFIG_INVENTORY_PATH
+    hash["ansible_ssh_user"] = ssh_auth.userid
+    hash["deployment"] = {}
+    hash["deployment"]["hosts"] = container_deployment_nodes.collect(&:to_ansible_config_format)
+    hash["deployment"]["roles"] = generate_roles
+    hash["version"] = version
+    hash["variant_version"] = kind.include?("origin") ? '1.2' : '3.2'
+    hash["variant"] = kind
+    # hash[:deployment][:roles][:master][:openshift_master_identity_providers] = []
+    # hash[:deployment][:roles][:master][:openshift_master_identity_providers] << identity_provider_auth.first.send(:to_ansible_config_format)
+    unless identity_provider_auth.first.htpassd_users.empty?
+      hash[:deployment][:roles][:master][:openshift_master_htpasswd_users] = {}
+      hash[:deployment][:roles][:master][:openshift_master_htpasswd_users] = identity_provider_auth.first.htpassd_users
+    end
+    hash.to_yaml.gsub("---\n", '')
+  end
+
+  def extract_public_ip_or_hostname(deployment_node)
+    return deployment_node.address if deployment_node.vm_id.nil?
+    vm = deployment_node.vm
+    hostname = vm.hostnames
+    if hostname.empty?
+      ips = vm.hardware.ipaddresses
+      ips.each do |ip|
+        return ip if public_ip? ip
+      end
+    end
+    hostname.first || ips.last
+  end
+
+  def public_ip?(ip)
```

Comment:
Can we try using https://github.com/ruby/ruby/blob/v2_2_5/test/socket/test_addrinfo.rb#L79-L96?

There is a way to check an `Addrinfo` object for private?, multicast?, and loopback?  Is a public ip just NOT one of these?  Maybe it's closer to what we want than to parse it ourselves.

See also: http://ruby-doc.org/stdlib-2.2.5/libdoc/socket/rdoc/Addrinfo.html#method-i-ipv4_private-3F